enhancements/baremetal/debuggability-of-baremetal-ipi-deployment.md

```diff
@@ -0,0 +1,256 @@
+
+### Goals
+
+- Any deployment that ends in an unsuccessful install must provide the
+  user clear and actionable information to diagnose the problem.
+
+### Non-Goals
+
+- Addressing the underlying causes of the failures is not the goal of
+  this enhancement.
+
+## Proposal
+
+Broadly, deployments fail due to problems encountered during these
+installation activities:
+
+  - Pre-bootstrap (image downloading, manifest creation, etc)
+  - Infrastructure automation (Terraform)
+  - Bootstrap
+  - Bare Metal Host Provisioning (Control Plane and Workers)
+  - Operator Deployment (i.e., those rolled out by CVO)
+
+We believe that since 4.5, pre-bootstrap errors are usually detected,
+and useful information is presented to the user about how to rectify the
+problem, so this enhancement request will focus on failures that occur
+from terraform onward.
+
+### Kinds of deployment failures
+
+#### Infrastructure Automation (Terraform)
+
+Bare metal IPI relies on terraform to provision a libvirt bootstrap
+virtual machine and the bare metal control plane hosts. We use
+`terraform-provider-libvirt` and `terraform-provider-ironic` to
+accomplish those goals.
+
+Both providers report failures when encountered, but there's usually
+little information provided to the user about what to do in the
+OpenShift context. Given a specific terraform error, the installer
+should provide specific information to the user about how to continue
+troubleshooting.
+
+#### Bootstrap Failures
+
+The bootstrap runs a couple of baremetal-specific services, including
+Ironic and a utility that populates hardware details for the control
+plane hosts.
+
+Bootstrap typically fails for baremetal when we can't download the
+machine-os image into our local HTTP cache.  Less common, but still
+sometimes seen are that services such as dnsmasq, mariadb, ironic-api,
+ironic-conductor, or ironic-inspector fail.
+
+Failures on bootstrap services rarely result in any indication to the
+user that something went wrong other than that there was a timeout.
+
+The installer has a feature for log gathering on bootstrap failure that
+does not work on baremetal. This should be the first priority, but even
+in this case a user still needs to look into an archive containing many
+logs to identify a failure.
+
+Ideally there would be some mechanism to identify and extract useful
+information and display it to the user.
+
+#### Bare Metal Host Provisioning
+
+Whether the control plane or worker nodes, provisioning of bare metal
+hosts can fail in the same ways, although the communication path to
+provide feedback is different in each case. For the control plane,
+information about failure is presented to the user via terraform. For
+workers, it is currently only shown on the `BareMetalHost` resource or
+by examining baremetal-operator logs. Failure to deploy a worker should
+be reflected by marking either the `machine-api-operator` or the future
+`cluster-baremetal-operator` degraded.
```

Comment:
We have to make sure this behavior is similar to behavior for other platforms. Should the MAO or CBO be marked degraded when 1 worker fails to deploy? What if that is an issue with the worker itself? How can we distinguish between errors in the resource being provisioned verses an issue in the control plane?

Reply:
These are all very good questions! It would be helpful to understand from someone in MAO (maybe @enxebre) how it works for other platforms, but my understanding is MAO doesn't go degraded from this case.

I think that MAO should show degraded if replicas are not met, or at the very least, if replicas are < 2, since we know we need 2 to get a working cluster on day 1 (unless controlplane is scheduable).

Perhaps cluster-operator-baremetal should also go degraded if provisioning fails, with more specific error messages bubbled up from baremetal-operator/ironic.

Reply:
After a discussion with the MAO team this is what I learnt. The MAO would go into a degraded state only when the pods that it is responsible for deploying, fail to come up. When a resource it manages, in this case a worker Machine, does not come up, the MAO does not go into a failed/degraded state for other platforms and should probably be the same for baremetal too.
When the initial deployment with 2 workers fails, then it should be considered an Installer error and we should provide the best/detailed errors message we can provide by bubbling up what we can get from BMO and/or Ironic. The MAO team believes that this is not a reason to put the operator in a degraded state. 
Since the baremetal platform is special, we could come up with a semantic on day 2 where if we notice a large number of worker failures (for example 90% of workers are not coming up), then the aggregated bad state could result in the operator being put in the degraded state. Currently, the operator does not have an aggregated view, so that needs to be added to the SLO at a future time.

Reply:
> After a discussion with the MAO team this is what I learnt. The MAO would go into a degraded state only when the pods that it is responsible for deploying, fail to come up. When a resource it manages, in this case a worker Machine, does not come up, the MAO does not go into a failed/degraded state for other platforms and should probably be the same for baremetal too.

Yea, it doesn't go to a degraded state for anyone and I think that's a mistake and what I'm proposing to change here.

> When the initial deployment with 2 workers fails, then it should be considered an Installer error and we should provide the best/detailed errors message we can provide by bubbling up what we can get from BMO and/or Ironic. The MAO team believes that this is not a reason to put the operator in a degraded state.

The main way the installer gets information about deployment success is largely through operator states via CVO, it would need a special case to count workers meeting the requested number of replicas.

If MAO can't provision machines, why is that not a degraded state of the operator? 

CC: @abhinavdahiya

Reply:
@stbenjam This docmentation https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource helped me understand the semantics behind different Operator states.

Reply:
>Failure to deploy a worker should
be reflected by marking either the `machine-api-operator` or the future
`cluster-baremetal-operator` degraded.

There are countless transient scenarios for "Failure to deploy a worker". This makes impractical putting a reasonable generic semantic on top of it. And so this makes worthless to let the overall operator going degraded in such a heterogeneous scenario.

Instead I believe the boundaries to signal the details of theses errors belong to individual machine resource conditions and any lower level resource. Just like we do for any other provider https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/apis/awsprovider/v1beta1/awsproviderstatus_types.go#L40

Then to communicate "Failure to deploy a worker" We already trigger alerts any time a machine has no node regardless of the failure details and regardless the provider. So each failure details can then be analysed in the format described above.

Beyond all the above, regardless of the failure details and based on the overall health of the cluster (e.g 99 out 100 machines has no node) we might decide our criteria for a semantic that represents a permanent global issue and choose to let the mao going degraded in that case. But that's a separate scoped discussion.

Reply:
> We already trigger alerts any time a machine has no node regardless of the failure details and regardless the provider.

These don't show up in the installer output, and as far as I know they do try to capture alerts. 

If you've ever done an install and ended up with a non-viable cluster due to insufficient workers, the UX is unacceptable. You get a report of a dozen failing operators and absolutely no indication it's because you don't have enough workers. @sdodson previously mentioned maybe we could do something in the installer about it (https://github.com/openshift/enhancements/pull/328#discussion_r427560374), which may help the problem I guess, but doesn't feel like the right solution to me as machine-api-operator, being the top-level operator for dealing with machines, should be signally clearly about the problem.

Reply:
>These don't show up in the installer output, and as far as I know they do try to capture alerts.

That'd be then a very a specific issue: "Installer output not capturing some existing alerts as expected".

>ended up with a non-viable cluster due to insufficient workers,

I agree. That scenario should be covered by my last statement in the previous comment.